src/Content/Text/BBCode.php

```diff
@@ -1872,56 +1854,42 @@ function (array $attributes, array $author_contact, $content, $is_quote_share) u
 
 			// Additionally, [pre] tags preserve spaces
 			$text = preg_replace_callback("/\[pre\](.*?)\[\/pre\]/ism", function ($match) {
-				return str_replace(' ', '&nbsp;', $match[1]);
+				return str_replace(' ', '&nbsp;', htmlentities($match[1], ENT_NOQUOTES,'UTF-8'));
 			}, $text);
 
+			// Add HTML new lines
+			$text = str_replace("\n", '<br>', $text);
+
 			return $text;
 		}); // Escaped code
 
 		$text = preg_replace_callback("#\[code(?:=([^\]]*))?\](.*?)\[\/code\]#ism",
 			function ($matches) {
 				if (strpos($matches[2], "\n") !== false) {
-					$return = '<pre><code class="language-' . trim($matches[1]) . '">' . htmlspecialchars(trim($matches[2], "\n\r"), ENT_NOQUOTES, 'UTF-8') . '</code></pre>';
+					$return = '<pre><code class="language-' . trim($matches[1]) . '">' . htmlentities(trim($matches[2], "\n\r"), ENT_NOQUOTES, 'UTF-8') . '</code></pre>';
 				} else {
-					$return = '<code>' . htmlspecialchars($matches[2], ENT_NOQUOTES, 'UTF-8') . '</code>';
+					$return = '<code>' . htmlentities($matches[2], ENT_NOQUOTES, 'UTF-8') . '</code>';
 				}
 
 				return $return;
 			},
 			$text
 		);
 
-		// Clean up the HTML by loading and saving the HTML with the DOM.
-		// Bad structured html can break a whole page.
-		// For performance reasons do it only with activated item cache or at export.
-		if (!$try_oembed || (get_itemcachepath() != '')) {
-			$doc = new DOMDocument();
-			$doc->preserveWhiteSpace = false;
-
-			$text = mb_convert_encoding($text, 'HTML-ENTITIES', "UTF-8");
-
-			$doctype = '<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">';
-			$encoding = '<?xml encoding="UTF-8">';
-			@$doc->loadHTML($encoding . $doctype . '<html><body>' . $text . '</body></html>');
-			$doc->encoding = 'UTF-8';
-			$text = $doc->saveHTML();
-			$text = str_replace(['<html><body>', '</body></html>', $doctype, $encoding], ['', '', '', ''], $text);
+		$config = \HTMLPurifier_HTML5Config::createDefault();
```

Comment:
Does the purifier also repairs broken HTML? Like adding missing closing elements and such stuff?

Reply:
It definitely repairs it, but I'm guessing it has a "pare down" approach rather than "the best display", so opening tags of expectedly closed elements probably would be discarded. Not sure what the `DOMDocument` approach is.

Reply:
`DOMDocument` behaves more like a regular browser. Means trying to make the best out of it without loosing content.

Are you running with it at the moment and you are partly comparing this with the same post from the original site?

Reply:
Not on my main node yet. I was expecting to use the RC to have more eyeballs on this feature change.

Reply:
Okay. So I hope that there will be no really ugly side effects ...